src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -332,77 +333,129 @@ public static unsafe void ClearWithoutReferences(ref byte b, nuint byteLength)
-            // mov instruction's base and index registers between loop iterations.
+            // Since references are always natural word-aligned, our "unaligned" writes below will
+            // always be natural word-aligned as well. Even if the full SIMD write is split across
+            // pages, no core will ever observe any reference as containing a torn address.
 
-            for (; pointerSizeLength >= 8; pointerSizeLength -= 8)
+            if (Vector.IsHardwareAccelerated && pointerSizeLength >= (uint)(Vector<byte>.Count / sizeof(IntPtr)))
             {
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -1) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -2) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -3) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -4) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -5) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -6) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -7) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -8) = default;
-            }
+                // We have enough data for at least one vectorized write.
+                // Perform that write now, potentially unaligned.
 
-            Debug.Assert(pointerSizeLength <= 7);
-
-            // The logic below works by trying to minimize the number of branches taken for any
-            // given range of lengths. For example, the lengths [ 4 .. 7 ] are handled by a single
-            // branch, [ 2 .. 3 ] are handled by a single branch, and [ 1 ] is handled by a single
-            // branch.
-            //
-            // We can write both forward and backward as a perf improvement. For example,
-            // the lengths [ 4 .. 7 ] can be handled by zeroing out the first four natural
-            // words and the last 3 natural words. In the best case (length = 7), there are
-            // no overlapping writes. In the worst case (length = 4), there are three
-            // overlapping writes near the middle of the buffer. In perf testing, the
-            // penalty for performing duplicate writes is less expensive than the penalty
-            // for complex branching.
-
-            if (pointerSizeLength >= 4)
-            {
-                goto Write4To7;
-            }
-            else if (pointerSizeLength >= 2)
-            {
-                goto Write2To3;
-            }
-            else if (pointerSizeLength > 0)
-            {
-                goto Write1;
+                Vector<byte> zero = default;
+                ref byte refDataAsBytes = ref Unsafe.As<IntPtr, byte>(ref ip);
+                Unsafe.WriteUnaligned(ref refDataAsBytes, zero);
```

Comment:
This assumes that `Unsafe.WriteUnaligned` is atomic for hardware accelerated vectors. It is not a safe assumption to make in general. It may the reason for the failing Mono tests.

Reply:
You are making assumption about the instruction that this is going to compile into.

The compiler (LLVM in this case) is free to compile this into byte-at-a-time copy, especially with optimization off.

Reply:
Also, IL interpretter is unlikely to execute this atomically either.

Reply:
If you would like to keep this, I think you may need `#define ATOMIC_UNALIGNED_VECTOR_OPERATIONS` that won't be set for MONO.

Reply:
`rep movsb` (`ERMSB`), for example, is a common optimization for simple copy loops with native compilers.

Reply:
>  It only needs to be atomic at a native word granularity. @tannergooding might know more, but I believe every retail or microprocessor in the market provides this guarantee.

Does the ARM spec actually guarantee this? I won't be surprised if ARM implementations with cheap memory controllers would not be required to guaratnee this.

Reply:
To clarify, you're talking about the `Vector.IsHardwareAccelerated` path, correct? The non-hardware-accelerated paths with this PR perform nuint-by-nuint writes, same as they did before.

Reply:
Yes, I am talking about `Vector.IsHardwareAccelerated` path. 

The code can be still running on IL interpretter even when `Vector.IsHardwareAccelerated` returns true.

Reply:
> Does the ARM spec actually guarantee this?

Yes for ARM64. ![image](https://user-images.githubusercontent.com/10487869/115435477-b952c680-a1be-11eb-9aaa-015b8aaf1c36.png)

I'm not 100% on how it handles ARM32 given that would be 4x 32-bit reads/writes. Probably need to dig further into the spec and the ARM32 specific sections.

Reply:
In particular:
> * Reads to SIMD and floating-point registers of a 128-bit value that is 64-bit aligned in memory are treated as a pair of single-copy atomic 64-bit reads.
> * Writes from SIMD and floating-point registers of a 128-bit value that is 64-bit aligned in memory are treated as a pair of single-copy atomic 64-bit writes.

Reply:
Also noting, I don't think this guarantees that something like LLVM won't rewrite the copy loop to be "more efficient". I'd imagine Mono is explicitly marking certain loops to prevent LLVM from doing such optimizations.

Reply:
What if we scrapped the `Vector.IsHardwareAccelerated` check and dropped down to raw SSE2 / ARM64 intrinsics instead? Presumably that wouldn't provide the opportunity for anybody to rewrite these methods in a non-compliant manner? Though I wonder if that would hurt perf since it'd require pinning (and the associated stack spill) to use the intrinsics.

Reply:
I imagine that depends on which instructions are being used and what optimizations LLVM is doing. It is conceivably "correct" in C/C++ code for it to recognize `__mm_load_si128` and `__mm_store_si128` as a "trivial copy loop" and to replace it with `ERMSB`.

(and https://godbolt.org/z/hsfE4nKnr shows that it does recognize and replace with `memcpy` 😄)